Source/Atomic/UI/UIFinderWindow.cpp

```diff
@@ -0,0 +1,618 @@
+        {
+            resultPath = "";  // were going back, give up file.
+            SetCurrentPath ( bookmarkPaths[selected]);
+            UpdateUiPath();
+            UpdateUiList();
+            UpdateUiResult();
+            bklist->SetValue(-1); // clear the select visuals
+        }
+        return true;
+    }
+    
+    if ( ev.type == EVENT_TYPE_CUSTOM && ev.target && ((uint32)ev.target->GetID()) == 5 ) // bookmarks TB context menu result
+    {
+        UI* ui = GetSubsystem<UI>();
+        if ( ev.special_key == tb::TB_KEY_DELETE ) // we wanna delete something
+        {
+            String myid;
+            ui->GetTBIDString(ev.target?((uint32)ev.target->GetID()) : 0, myid);
+            UISelectList *bklist = static_cast<UISelectList *>(GetBookmarksWidget());
+            if (bklist)
+            {
+                int myindex = bklist->FindId ( (uint32)ev.ref_id );
+                if ( myindex >= 0 )
+                {
+                    DeleteBookmark(myindex);
+                }
+            }
+        }
+        return true;
+    }
+
+    if ( ev.type == EVENT_TYPE_CLICK && ev.target && ((uint32)ev.target->GetID()) == 6 ) // clicked dirfiles list
+    {
+        UISelectList *filelist = static_cast<UISelectList *>(GetFilesWidget());
+        ComposePath( filelist->GetSelectedItemString() );
+        return true;
+    }
+
+    if ( ev.type == EVENT_TYPE_CLICK && (ev.ref_id == 8 || ev.ref_id == 9) ) // clicked 8 = ok button, 9 = cancel button
+    {
+        UI* ui = GetSubsystem<UI>();
+        VariantMap eventData;
+
+        String title = "FinderWindow";
+        TBStr tbtext;
+        if(  widget_ && (TBWindow*)widget_->GetText(tbtext) )
+            title = tbtext.CStr();
+
+        eventData[UIFinderComplete::P_TITLE] = title;
+        eventData[UIFinderComplete::P_SELECTED] = "";
+        eventData[UIFinderComplete::P_REASON] = "CANCEL";
+
+        if (ev.ref_id == 8) // ok button was pressed, otherwise it was cancel button
+        {
+            eventData[UIFinderComplete::P_REASON] = "OK";
+            if ( finderMode == 0 ) // finding a file
+            {   // get from widget, in case the user had been typing.
+                UIWidget *ewidget = GetResultWidget(); 
+                if( ewidget) eventData[UIFinderComplete::P_SELECTED] = ewidget->GetText();
+            }
+            else  // finding a folder
+            {
+                UIWidget *cwidget = GetPathWidget();
+                if( cwidget) eventData[UIFinderComplete::P_SELECTED] = cwidget->GetText();
+            }
+        }
+
+        ConvertEvent(this, ui->WrapWidget(ev.target), ev, eventData);
+        SendEvent(E_UIFINDERCOMPLETE, eventData);
+
+        if (eventData[WidgetEvent::P_HANDLED].GetBool())
+           return true;
+    }
+
+    return UIWindow::OnEvent(ev);
+}
+
+void UIFinderWindow::HandleCreateBookmark(StringHash eventType, VariantMap& eventData)
+{
+    String Title = eventData["Title"].GetString();
```

Comment:
Just a FYI, these can be const String& to avoid a copy from eventData variant map